File: 2°periodo/AEDS2/tps/tp04/java/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.io.File;

class Game {
    //atributos
    private int id;
    private String name;
    private String releaseDate;
    private int estimatedOwners;
    private float price;
    private String[] supportedLanguages;
    private int metacriticScore;
    private float userScore;
    private int achievements;
    private String[] publishers;
    private String[] developers;
    private String[] categories;
    private String[] genres;
    private String[] tags;

    public Game() {}

    // sets e get
    public void setId(int id) { this.id = id; }
    public void setName(String name) { this.name = name; }
    public void setReleaseDate(String releaseDate) { this.releaseDate = releaseDate; }
    public void setEstimatedOwners(int estimatedOwners) { this.estimatedOwners = estimatedOwners; }
    public void setPrice(float price) { this.price = price; }
    public void setSupportedLanguages(String[] supportedLanguages) { this.supportedLanguages = supportedLanguages; }
    public void setMetacriticScore(int metacriticScore){ this.metacriticScore = metacriticScore; }
    public void setUserScore(float userScore){ this.userScore = userScore; }
    public void setAchievements(int achievements){ this.achievements = achievements; }
    public void setPublishers(String[] publishers) { this.publishers = publishers; }
    public void setDevelopers(String[] developers) { this.developers = developers; }
    public void setCategories(String[] categories) { this.categories = categories; }
    public void setGenres(String[] genres) { this.genres = genres; }
    public void setTags(String[] tags) { this.tags = tags; }
    public int getId() { return id; }


    
      //Pega todos os dados de um jogo e os imprime na tela em uma única linha
     
    public void imprimir() {
        System.out.print("=> " + id + " ## " + name + " ## " + releaseDate + " ## " + estimatedOwners + " ## " + price + " ## [");
        for (int i = 0; i < supportedLanguages.length; i++) {
            System.out.print(supportedLanguages[i]);
            if (i < supportedLanguages.length - 1) System.out.print(", ");
        }
        System.out.print("] ## " + metacriticScore + " ## " + userScore + " ## " + achievements + " ## [");
        for (int i = 0; i < publishers.length; i++) {
            System.out.print(publishers[i]);
            if (i < publishers.length - 1) System.out.print(", ");
        }
        System.out.print("] ## [");
        for (int i = 0; i < developers.length; i++) {
            System.out.print(developers[i]);
            if (i < developers.length - 1) System.out.print(", ");
        }
        System.out.print("] ## [");
        for (int i = 0; i < categories.length; i++) {
            System.out.print(categories[i]);
            if (i < categories.length - 1) System.out.print(", ");
        }
        System.out.print("] ## [");
        for (int i = 0; i < genres.length; i++) {
            System.out.print(genres[i]);
            if (i < genres.length - 1) System.out.print(", ");
        }
        System.out.print("] ## [");
        for (int i = 0; i < tags.length; i++) {
            System.out.print(tags[i]);
            if (i < tags.length - 1) System.out.print(", ");
        }
        // para pular para a próxima linha após imprimir tudo.
        System.out.println("] ##");
    }
}

public class Main {
    /*
      Um método auxiliar para fazer o trabalho do "split".
      Ele pega um texto (ex: "A,B,C") e um separador (ex: ',') e retorna
      um array de strings (ex: ["A", "B", "C"]).
     */
    private static String[] separarStringInterna(String texto, char separador) {
        if (texto == null || texto.isEmpty()) return new String[0];
        List<String> resultado = new ArrayList<>();
        String itemAtual = "";
        for (int i = 0; i < texto.length(); i++) {
            char c = texto.charAt(i);
            if (c == separador) {
                resultado.add(itemAtual);
                itemAtual = "";
                if (i + 1 < texto.length() && texto.charAt(i + 1) == ' ') i++;
            } else {
                itemAtual += c;
            }
        }
        resultado.add(itemAtual);
        String[] arrayResultado = new String[resultado.size()];
        for(int i=0; i < resultado.size(); i++) arrayResultado[i] = resultado.get(i);
        return arrayResultado;
    }

    public static void formatar(Game game, String linha) {
        
        /* percorre a linha do CSV
        caractere por caractere. A variável `dentroDeAspas` garante que ele não
         separe campos por vírgulas que estão dentro de aspas (como em uma data).*/
        List<String> campos = new ArrayList<>();
        String campoAtual = "";
        boolean dentroDeAspas = false;
    
        for (int i = 0; i < linha.length(); i++) {
            char c = linha.charAt(i);
            if (c == '"') {
                dentroDeAspas = !dentroDeAspas;
            } else if (c == ',' && !dentroDeAspas) {
                campos.add(campoAtual);
                campoAtual = "";
            } else {
                campoAtual += c;
            }
        }
        campos.add(campoAtual);
    
        
        game.setId(Integer.parseInt(campos.get(0)));
        game.setName(campos.get(1));
        
        
        String dataOriginal = campos.get(2);
        String dataSemVirgula = "";
        for(int i = 0; i < dataOriginal.length(); i++){
            if(dataOriginal.charAt(i) != ',') dataSemVirgula += dataOriginal.charAt(i);
        }
        String[] dataPartes = separarStringInterna(dataSemVirgula, ' ');
        
        
        String diaStr = "01", mesStr = "Jan", anoStr = "1970";

        if (dataPartes.length >= 3) { 
            mesStr = dataPartes[0];
            diaStr = dataPartes[1];
            anoStr = dataPartes[2];
        } else if (dataPartes.length == 2) { 
            mesStr = dataPartes[0];
            anoStr = dataPartes[1];
            diaStr = "01"; 
        } 
        
        
        if (diaStr.length() == 1) {
            diaStr = "0" + diaStr;
        }

        String mesNum = "";
        switch (mesStr) {
            case "Jan": mesNum = "01"; break; case "Feb": mesNum = "02"; break;
            case "Mar": mesNum = "03"; break; case "Apr": mesNum = "04"; break;
            case "May": mesNum = "05"; break; case "Jun": mesNum = "06"; break;
            case "Jul": mesNum = "07"; break; case "Aug": mesNum = "08"; break;
            case "Sep": mesNum = "09"; break; case "Oct": mesNum = "10"; break;
            case "Nov": mesNum = "11"; break; case "Dec": mesNum = "12"; break;
            default: mesNum = "00"; break;
        }
        game.setReleaseDate(diaStr + "/" + mesNum + "/" + anoStr);
        
        
        game.setEstimatedOwners(Integer.parseInt(campos.get(3)));
        game.setPrice(Float.parseFloat(campos.get(4)));
        String linguagensStr = campos.get(5);
        String linguagensLimpas = "";
        for(int i = 0; i < linguagensStr.length(); i++){
            char c = linguagensStr.charAt(i);
            if(c != '[' && c != ']' && c != '\'') linguagensLimpas += c;
        }
        game.setSupportedLanguages(separarStringInterna(linguagensLimpas, ','));
        game.setMetacriticScore(Integer.parseInt(campos.get(6)));
        game.setUserScore(Float.parseFloat(campos.get(7)));
        game.setAchievements(Integer.parseInt(campos.get(8)));
        game.setPublishers(separarStringInterna(campos.get(9), ','));
        game.setDevelopers(separarStringInterna(campos.get(10), ','));
        game.setCategories(separarStringInterna(campos.get(11), ','));
        game.setGenres(separarStringInterna(campos.get(12), ','));
        game.setTags(separarStringInterna(campos.get(13), ','));
    }

    public static void main(String[] args) throws Exception {
        
        // Lê o arquivo "games.csv", chama o 'formatar' para cada linha
        // e guarda todos os objetos 'Game' em uma lista.
        File arquivo = new File("/tmp/games.csv");
        List<Game> gamesList = new ArrayList<>();
        try (Scanner scFile = new Scanner(arquivo)) {
            scFile.nextLine(); // Pula o cabeçalho
            while (scFile.hasNextLine()) {
                String linha = scFile.nextLine();
                if (linha.isEmpty()) continue;
                Game novoGame = new Game();
                formatar(novoGame, linha);
                gamesList.add(novoGame);
            }
        }
    
        
        try (Scanner sc = new Scanner(System.in)) {
            String entrada;
            
            while (sc.hasNext()) {
                entrada = sc.next(); 
    
                // Se o usuário digitar "FIM", o programa para.
                if (entrada.equalsIgnoreCase("FIM")) {
                    break;
                }
    
                // Converte a entrada para número e procura o jogo na lista.
                int idParaBuscar = Integer.parseInt(entrada);
                for (Game game : gamesList) {
                    if (game.getId() == idParaBuscar) {
                        game.imprimir();
                        break; 
                    }
                }
            }
        } 
    } 
}

File: 2°periodo/AEDS2/tps/tp04/c/Main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <strings.h> 
#include <ctype.h> 

typedef struct {
    //atributos
    int id;
    char* name;
    char* releaseDate;
    int estimatedOwners;
    float price;
    char** supportedLanguages;
    int supportedLanguages_count;
    int metacriticScore;
    float userScore;
    int achievements;
    char** publishers;
    int publishers_count;
    char** developers;
    int developers_count;
    char** categories;
    int categories_count;
    char** genres;
    int genres_count;
    char** tags;
    int tags_count;
} Game;

//funções auxiliares
void formatar(Game* game, const char* linha);
void imprimir(const Game* game);
void free_game_memory(Game* game);
char** separarStringInterna(const char* texto, char separador, int* count);
char* trim(char* str);

//Pega todos os dados de um jogo e os imprime na tela em uma única linha
void imprimir(const Game* game) {
    printf("=> %d ## %s ## %s ## %d ## ", game->id, game->name, game->releaseDate, game->estimatedOwners);

    if (game->price == 0.0f) {
        printf("%.1f", game->price); 
    } else {
        printf("%g", game->price); 
    }
   
    printf(" ## [");
    for (int i = 0; i < game->supportedLanguages_count; i++) {
        printf("%s", game->supportedLanguages[i]);
        if (i < game->supportedLanguages_count - 1) printf(", ");
    }
    printf("] ## %d ## %.1f ## %d ## [", game->metacriticScore, game->userScore, game->achievements);
    for (int i = 0; i < game->publishers_count; i++) {
        printf("%s", game->publishers[i]);
        if (i < game->publishers_count - 1) printf(", ");
    }
    printf("] ## [");
    for (int i = 0; i < game->developers_count; i++) {
        printf("%s", game->developers[i]);
        if (i < game->developers_count - 1) printf(", ");
    }
    printf("] ## [");
    for (int i = 0; i < game->categories_count; i++) {
        printf("%s", game->categories[i]);
        if (i < game->categories_count - 1) printf(", ");
    }
    printf("] ## [");
    for (int i = 0; i < game->genres_count; i++) {
        printf("%s", game->genres[i]);
        if (i < game->genres_count - 1) printf(", ");
    }
    printf("] ## [");
    for (int i = 0; i < game->tags_count; i++) {
        printf("%s", game->tags[i]);
        if (i < game->tags_count - 1) printf(", ");
    }
    // para pular para a próxima linha após imprimir tudo.
    printf("] ##\n");
}


void formatar(Game* game, const char* linha) {
    /* percorre a linha do CSV
    caractere por caractere. A variável `dentroDeAspas` garante que ele não
     separe campos por vírgulas que estão dentro de aspas (como em uma data).*/
    char** campos = NULL;
    int campos_count = 0;
    char* campoAtual = malloc(2048);
    int campoAtual_len = 0;
    bool dentroDeAspas = false;

    for (int i = 0; linha[i] != '\0'; i++) {
        char c = linha[i];
        if (c == '"') {
            dentroDeAspas = !dentroDeAspas;
        } else if (c == ',' && !dentroDeAspas) {
            campoAtual[campoAtual_len] = '\0';
            campos_count++;
            campos = realloc(campos, campos_count * sizeof(char*));
            campos[campos_count - 1] = strdup(campoAtual);
            campoAtual_len = 0;
        } else {
            if (campoAtual_len < 2047) {
                campoAtual[campoAtual_len++] = c;
            }
        }
    }
    campoAtual[campoAtual_len] = '\0';
    campos_count++;
    campos = realloc(campos, campos_count * sizeof(char*));
    campos[campos_count - 1] = strdup(campoAtual);
    free(campoAtual);

    for (int i = 0; i < campos_count; i++) {
        trim(campos[i]);
    }
    
    game->id = atoi(campos[0]);
    game->name = strdup(campos[1]);

    char dataSemVirgula[100] = {0};
    for (int i = 0, j = 0; campos[2][i] != '\0'; i++) {
        if (campos[2][i] != ',') {
            dataSemVirgula[j++] = campos[2][i];
        }
    }
    int dataPartes_count = 0;
    char** dataPartes = separarStringInterna(dataSemVirgula, ' ', &dataPartes_count);
    char *diaStr = "01", *mesStr = "Jan", *anoStr = "1970";

    if (dataPartes_count >= 3) {
        mesStr = dataPartes[0]; diaStr = dataPartes[1]; anoStr = dataPartes[2];
    } else if (dataPartes_count == 2) {
        mesStr = dataPartes[0]; anoStr = dataPartes[1]; diaStr = "01";
    }
    
    char diaFinal[3];
    if (strlen(diaStr) == 1) sprintf(diaFinal, "0%s", diaStr);
    else { strncpy(diaFinal, diaStr, 2); diaFinal[2] = '\0'; }

    char mesNum[3] = "00";
    if (strcmp(mesStr, "Jan") == 0) strcpy(mesNum, "01"); else if (strcmp(mesStr, "Feb") == 0) strcpy(mesNum, "02");
    else if (strcmp(mesStr, "Mar") == 0) strcpy(mesNum, "03"); else if (strcmp(mesStr, "Apr") == 0) strcpy(mesNum, "04");
    else if (strcmp(mesStr, "May") == 0) strcpy(mesNum, "05"); else if (strcmp(mesStr, "Jun") == 0) strcpy(mesNum, "06");
    else if (strcmp(mesStr, "Jul") == 0) strcpy(mesNum, "07"); else if (strcmp(mesStr, "Aug") == 0) strcpy(mesNum, "08");
    else if (strcmp(mesStr, "Sep") == 0) strcpy(mesNum, "09"); else if (strcmp(mesStr, "Oct") == 0) strcpy(mesNum, "10");
    else if (strcmp(mesStr, "Nov") == 0) strcpy(mesNum, "11"); else if (strcmp(mesStr, "Dec") == 0) strcpy(mesNum, "12");

    game->releaseDate = malloc(11);
    sprintf(game->releaseDate, "%s/%s/%s", diaFinal, mesNum, anoStr);

    game->estimatedOwners = atoi(campos[3]);
    game->price = atof(campos[4]);
    
    char linguagensLimpas[1024] = {0};
    for (int i = 0, j = 0; campos[5][i] != '\0'; i++) {
        if (campos[5][i] != '[' && campos[5][i] != ']' && campos[5][i] != '\'') {
            linguagensLimpas[j++] = campos[5][i];
        }
    }
    game->supportedLanguages = separarStringInterna(linguagensLimpas, ',', &game->supportedLanguages_count);
    
    game->metacriticScore = atoi(campos[6]);
    game->userScore = atof(campos[7]);
    game->achievements = atoi(campos[8]);
    game->publishers = separarStringInterna(campos[9], ',', &game->publishers_count);
    game->developers = separarStringInterna(campos[10], ',', &game->developers_count);
    game->categories = separarStringInterna(campos[11], ',', &game->categories_count);
    game->genres = separarStringInterna(campos[12], ',', &game->genres_count);
    game->tags = separarStringInterna(campos[13], ',', &game->tags_count);
    
    for (int i = 0; i < campos_count; i++) free(campos[i]);
    free(campos);
    for (int i = 0; i < dataPartes_count; i++) free(dataPartes[i]);
    free(dataPartes);
}


int main(void) {
    // Lê o arquivo "games.csv", chama o 'formatar' para cada linha
    // e guarda todos os objetos 'Game' em uma lista.
    FILE* arquivo = fopen("/tmp/games.csv", "r");
    if (arquivo == NULL) {
        printf("ERRO: Nao foi possivel abrir o arquivo 'games.csv'\n");
        return 1;
    }

    int games_capacity = 40000;
    int games_size = 0;
    Game* gamesList = malloc(games_capacity * sizeof(Game));
    char linha[8192]; 

    fscanf(arquivo, "%[^\n]%*c", linha); // Pula o cabeçalho

    while (fscanf(arquivo, " %[^\n]%*c", linha) != EOF) {
        if (games_size >= games_capacity) {
            games_capacity *= 2;
            gamesList = realloc(gamesList, games_capacity * sizeof(Game));
        }
        formatar(&gamesList[games_size], linha);
        games_size++;
    }
    fclose(arquivo);

    char entrada[100];
    while (scanf("%s", entrada) != EOF) {
        // Se o usuário digitar "FIM", o programa para.
        if (strcasecmp(entrada, "FIM") == 0) {
            break;
        }

        // Converte a entrada para número e procura o jogo na lista.
        int idParaBuscar = atoi(entrada);
        bool achou = false;
        for (int i = 0; i < games_size; i++) {
            if (gamesList[i].id == idParaBuscar) {
                imprimir(&gamesList[i]);
                achou = true;
                break;
            }
        }
    }
    
    for (int i = 0; i < games_size; i++) {
        free_game_memory(&gamesList[i]);
    }
    free(gamesList);

    return 0;
}


char* trim(char* str) {
    if (str == NULL) return NULL;
    char* end;
    while (isspace((unsigned char)*str)) str++;
    if (*str == 0) return str;
    end = str + strlen(str) - 1;
    while (end > str && isspace((unsigned char)*end)) end--;
    end[1] = '\0';
    return str;
}


void free_game_memory(Game* game) {
    free(game->name);
    free(game->releaseDate);
    for (int i = 0; i < game->supportedLanguages_count; i++) free(game->supportedLanguages[i]);
    free(game->supportedLanguages);
    for (int i = 0; i < game->publishers_count; i++) free(game->publishers[i]);
    free(game->publishers);
    for (int i = 0; i < game->developers_count; i++) free(game->developers[i]);
    free(game->developers);
    for (int i = 0; i < game->categories_count; i++) free(game->categories[i]);
    free(game->categories);
    for (int i = 0; i < game->genres_count; i++) free(game->genres[i]);
    free(game->genres);
    for (int i = 0; i < game->tags_count; i++) free(game->tags[i]);
    free(game->tags);
}

/*
  Um método auxiliar para fazer o trabalho do "split".
  Ele pega um texto (ex: "A,B,C") e um separador (ex: ',') e retorna
  um array de strings (ex: ["A", "B", "C"]).
 */
char** separarStringInterna(const char* texto, char separador, int* count) {
    *count = 0;
    if (texto == NULL || texto[0] == '\0') {
        return NULL;
    }

    int temp_count = 1;
    for (int i = 0; texto[i] != '\0'; i++) {
        if (texto[i] == separador) temp_count++;
    }
    
    char** resultado = malloc(temp_count * sizeof(char*));
    char itemAtual[1024] = {0};
    int itemAtual_len = 0;
    int resultado_idx = 0;

    for (int i = 0; texto[i] != '\0'; i++) {
        if (texto[i] == separador) {
            itemAtual[itemAtual_len] = '\0';
            resultado[resultado_idx++] = strdup(trim(itemAtual));
            itemAtual_len = 0;
        } else {
            if (itemAtual_len < 1023) {
                itemAtual[itemAtual_len++] = texto[i];
            }
        }
    }
    itemAtual[itemAtual_len] = '\0';
    resultado[resultado_idx++] = strdup(trim(itemAtual));
    
    *count = resultado_idx;
    return resultado;
}